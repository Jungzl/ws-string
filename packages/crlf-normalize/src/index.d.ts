import { ITSAndTypeAndStringLiteral, ITSTypeAndStringLiteral } from 'ts-type/lib/helper/string';
export declare const enum EnumLineBreak {
    CR = "\r",
    CRLF = "\r\n",
    LF = "\n"
}
declare const CR: EnumLineBreak.CR;
declare const CRLF: EnumLineBreak.CRLF;
declare const LF: EnumLineBreak.LF;
declare const R_CRLF: RegExp;
export { CR, CRLF, LF, R_CRLF };
export type ILineBreak = ITSTypeAndStringLiteral<EnumLineBreak>;
export type ILineBreakInput = ITSAndTypeAndStringLiteral<EnumLineBreak>;
export interface IOptions {
    disable?: {
        lf?: boolean;
        crlf?: boolean;
        cr?: boolean;
    };
}
export declare function crlf(text: string, newline?: ILineBreakInput): string;
export declare function chkcrlf(text: string, options?: IOptions): {
    readonly lf: boolean;
    readonly crlf: boolean;
    readonly cr: boolean;
};
export declare function detectLineBreak(text: string, options?: IOptions): EnumLineBreak;
export declare function isCRLF(newline: string): newline is EnumLineBreak.CRLF;
export declare function isLF(newline: string): newline is EnumLineBreak.LF;
export declare function isCR(newline: string): newline is EnumLineBreak.CR;
export declare function lineSplit(text: string): string[];
export declare function crlf_unicode_normalize(text: string, newline?: ILineBreakInput): string;
export default crlf;
