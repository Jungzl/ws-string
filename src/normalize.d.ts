/**
 * Created by user on 2018/1/19/019.
 */
export declare const LF = "\n";
export declare function normalize(txt: any): string;
export default normalize;
