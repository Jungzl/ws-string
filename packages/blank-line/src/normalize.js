"use strict";
/**
 * Created by user on 2018/1/19/019.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LF = "\n";
function normalize(txt) {
    return String(txt)
        .replace(/\r\n|\r(?!\n)|\n/g, exports.LF)
        .replace(/\uFEFF/g, '')
        .replace(/[  \xA0]/g, ' ')
        .replace(/[ \t　\xA0\u3000]+\n/g, '\n')
        .replace(/^\n+|[\s　\xA0\u3000]+$/g, '');
}
exports.normalize = normalize;
exports.default = normalize;
//export default exports;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoibm9ybWFsaXplLmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsibm9ybWFsaXplLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7QUFBQTs7R0FFRzs7QUFFVSxRQUFBLEVBQUUsR0FBRyxJQUFJLENBQUM7QUFFdkIsU0FBZ0IsU0FBUyxDQUFDLEdBQUc7SUFFNUIsT0FBTyxNQUFNLENBQUMsR0FBRyxDQUFDO1NBQ2hCLE9BQU8sQ0FBQyxtQkFBbUIsRUFBRSxVQUFFLENBQUM7U0FDaEMsT0FBTyxDQUFDLFNBQVMsRUFBRSxFQUFFLENBQUM7U0FDdEIsT0FBTyxDQUFDLFdBQVcsRUFBRSxHQUFHLENBQUM7U0FDekIsT0FBTyxDQUFDLHNCQUFzQixFQUFFLElBQUksQ0FBQztTQUNyQyxPQUFPLENBQUMseUJBQXlCLEVBQUUsRUFBRSxDQUFDLENBQ3ZDO0FBQ0YsQ0FBQztBQVRELDhCQVNDO0FBRUQsa0JBQWUsU0FBUyxDQUFDO0FBQ3pCLHlCQUF5QiIsInNvdXJjZXNDb250ZW50IjpbIi8qKlxuICogQ3JlYXRlZCBieSB1c2VyIG9uIDIwMTgvMS8xOS8wMTkuXG4gKi9cblxuZXhwb3J0IGNvbnN0IExGID0gXCJcXG5cIjtcblxuZXhwb3J0IGZ1bmN0aW9uIG5vcm1hbGl6ZSh0eHQpXG57XG5cdHJldHVybiBTdHJpbmcodHh0KVxuXHRcdC5yZXBsYWNlKC9cXHJcXG58XFxyKD8hXFxuKXxcXG4vZywgTEYpXG5cdFx0LnJlcGxhY2UoL1xcdUZFRkYvZywgJycpXG5cdFx0LnJlcGxhY2UoL1sgwqBcXHhBMF0vZywgJyAnKVxuXHRcdC5yZXBsYWNlKC9bIFxcdOOAgFxceEEwXFx1MzAwMF0rXFxuL2csICdcXG4nKVxuXHRcdC5yZXBsYWNlKC9eXFxuK3xbXFxz44CAXFx4QTBcXHUzMDAwXSskL2csICcnKVxuXHQ7XG59XG5cbmV4cG9ydCBkZWZhdWx0IG5vcm1hbGl6ZTtcbi8vZXhwb3J0IGRlZmF1bHQgZXhwb3J0cztcbiJdfQ==