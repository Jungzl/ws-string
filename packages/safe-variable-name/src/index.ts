import camelCase from 'camelcase';
import { isReservedOrBuiltinsLC } from 'reserved2';

export function _safeVariableName(input: string)
{
	return camelCase(
		input.replace(/([^\w\-_ ])/ig, '-')
		, { pascalCase: true, preserveConsecutiveUppercase: true, locale: 'en-US' })
		.replace(/((^[^a-zA-Z]+)|[^\w.-])|([^a-zA-Z0-9]+$)/g, '');
}

export function safeVariableName(input: string)
{
	let name = _safeVariableName(input);

	if (isReservedOrBuiltinsLC(name.toLowerCase()))
	{
		name = '_' + name;
	}

	return name
}

export default safeVariableName
