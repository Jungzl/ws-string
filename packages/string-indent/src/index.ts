const REGEX_MATCH_INDENT = /^([ \t]*)((?:[\S\r\n].*|)$)/;
const REGEX_MATCH_INDENT_NO_BREAK_SPACE = /^([ \t\xa0]*)((?:[\S\r\n].*|)$)/;

export interface IOptions
{
	/**
	 * no-break space
	 * \xa0
	 */
	includeNoBreakSpace?: boolean,
}

export function detectIndentLine(input: string, options?: IOptions)
{
	const m = (options?.includeNoBreakSpace ? REGEX_MATCH_INDENT_NO_BREAK_SPACE : REGEX_MATCH_INDENT).exec(input);
	return {
		input,
		indent: m[1],
		body: m[2],
		bool: m[1].length > 0,
	}
}

export default detectIndentLine
