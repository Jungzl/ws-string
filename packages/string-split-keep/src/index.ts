"use strict"

export function stringSplitWithLimit(str: string, separator?: string, limit?: number): string[]
{
	if (!limit || separator === undefined) return str.split(separator, limit)
	if (!separator.length) return limit > 0 ? _splitStart_noSep(str, limit) : _splitEnd_noSep(str, -limit)
	return limit > 0 ? _splitStart(str, separator, limit, 0, []) : _splitEnd(str, separator, -limit, str.length, [])
}

export default stringSplitWithLimit

export function _splitStart_noSep(str: string, lim: number)
{
	const ret: string[] = [];
	lim = Math.min(lim, str.length) - 1
	for (let i = 0; i < lim; ++i) ret.push(str[i])
	ret.push(str.slice(lim))
	return ret
}

export function _splitEnd_noSep(str: string, lim: number)
{
	const ret: string[] = [];
	lim = Math.min(lim, str.length)
	const firstSlice = str.length - lim + 1;
	ret.push(str.slice(0, firstSlice))
	for (let i = firstSlice; i < str.length; ++i) ret.push(str[i])
	return ret
}

export function _splitStart(str: string, sep: string, lim: number, cur: number, acc: string[]): string[]
{
	const index = str.indexOf(sep, cur);
	if (index == -1 || acc.length + 1 == lim)
	{
		acc.push(str.slice(cur))
		return acc
	}
	acc.push(str.slice(cur, index))
	return _splitStart(str, sep, lim, index + sep.length, acc)
}

export function _splitEnd(str: string, sep: string, lim: number, cur: number, acc: string[]): string[]
{
	const index = str.lastIndexOf(sep, cur);
	if (cur == -1 || index == -1 || acc.length + 1 == lim)
	{
		acc.unshift(str.slice(0, cur + 1))
		return acc
	}
	acc.unshift(str.slice(index + sep.length, cur + 1))
	return _splitEnd(str, sep, lim, index - 1, acc)
}
