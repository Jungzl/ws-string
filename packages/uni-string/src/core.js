"use strict";
/**
 * Created by user on 2018/3/16/016.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
exports.__esModule = true;
exports.UString = exports.STRING_PROTOTYPE = void 0;
var runes2_1 = require("runes2");
var es6_class_prototype_1 = require("es6-class-prototype");
exports.STRING_PROTOTYPE = Object.getOwnPropertyNames(String.prototype);
var UString = /** @class */ (function (_super) {
    __extends(UString, _super);
    function UString(str) {
        var argv = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            argv[_i - 1] = arguments[_i];
        }
        var _this = _super.call(this, str) || this;
        _this._arr = null;
        var d = Object.getOwnPropertyDescriptor(_this, '_arr');
        Object.defineProperty(_this, '_arr', Object.assign(d, {
            configurable: true,
            enumerable: false,
        }));
        return _this;
    }
    UString.prototype[Symbol.iterator] = function () {
        return UString.toArray(this)[Symbol.iterator]();
    };
    UString.isString = function (str) {
        return (typeof str == 'string' || str instanceof String);
    };
    UString.toArray = function (str) {
        if (str instanceof UString) {
            return str.toArray();
        }
        return runes2_1["default"](String(str));
    };
    UString.prototype.toArray = function () {
        if (!this._arr) {
            this._arr = runes2_1["default"](String(this));
        }
        return this._arr;
    };
    UString.prototype.split = function (separator, limit) {
        var ret;
        var str = String(this);
        if (separator === '') {
            ret = UString.toArray(this);
            if (typeof limit != 'undefined') {
                ret = ret.slice(0, limit);
            }
        }
        else {
            ret = String.prototype.split.call(str, separator, limit);
        }
        return ret;
    };
    UString.prototype.substr = function (start, length) {
        return UString.toArray(this).slice(start).slice(0, length).join('');
    };
    UString.prototype.substring = function (start, indexEnd) {
        var _a;
        if (Number.isNaN(start) || start < 0) {
            start = 0;
        }
        if (typeof indexEnd == 'number') {
            if (Number.isNaN(indexEnd) || indexEnd < 0) {
                indexEnd = 0;
            }
            if (start > indexEnd) {
                _a = [indexEnd, start], start = _a[0], indexEnd = _a[1];
            }
        }
        return this.slice(start, indexEnd);
    };
    UString.prototype.indexOf = function (search, start) {
        if (start === void 0) { start = 0; }
        var a = UString.toArray(this);
        start = Math.max(0, Math.min(a.length, start));
        search = String(search);
        if (search === '') {
            return start;
        }
        a = a.slice(start);
        var s = UString.toArray(search);
        var i = 0;
        var j = 0;
        var se = s[s.length - 1];
        do {
            i = a.indexOf(s[0], j);
            if (i != -1) {
                if (a.slice(i, i + s.length).join('') == search) {
                    return start + i;
                }
                j = i;
                if (s.length > 1) {
                    i = a.indexOf(se, i + 1);
                    j = i - s.length;
                }
            }
            j++;
        } while (i != -1 && j < a.length);
        return -1;
    };
    UString.prototype.includes = function (search, start) {
        if (start === void 0) { start = 0; }
        return UString.toArray(this).slice(start).join('').indexOf(search) !== -1;
    };
    UString.prototype.slice = function (start, indexEnd) {
        return UString.toArray(this).slice(start, indexEnd).join('');
    };
    UString.prototype.charAt = function (index) {
        return this.substr(index, 1);
    };
    UString.prototype.startsWith = function (search, pos) {
        return this.substr(!pos || pos < 0 ? 0 : +pos, search.length)
            .indexOf(search) == 0;
    };
    UString.prototype.endsWith = function (search, length) {
        var a = UString.toArray(this);
        var s = UString.toArray(search);
        if (length === undefined || length > a.length) {
            length = a.length;
        }
        return this.substring(length - s.length, length) === search;
    };
    UString.prototype.padEnd = function (targetLength, padString) {
        targetLength = targetLength >> 0; //floor if number or convert non-number to 0;
        padString = String((typeof padString !== 'undefined' ? padString : ' '));
        var str = this.split('');
        var pad = this.split.call(padString, '');
        if (str.length > targetLength) {
            return String(this);
        }
        else {
            targetLength = targetLength - str.length;
            if (targetLength > pad.length) {
                padString += padString.repeat(targetLength / pad.length); //append to original to ensure we are longer than needed
                // @ts-ignore
                pad = new UString(padString);
            }
            return String(this) + pad.slice(0, targetLength);
        }
    };
    UString.prototype.padStart = function (targetLength, padString) {
        targetLength = targetLength >> 0; //truncate if number or convert non-number to 0;
        padString = String((typeof padString !== 'undefined' ? padString : ' '));
        var str = this.split('');
        var pad = this.split.call(padString, '');
        if (str.length > targetLength) {
            return String(this);
        }
        else {
            targetLength = targetLength - str.length;
            if (targetLength > pad.length) {
                padString += padString.repeat(targetLength / pad.length); //append to original to ensure we are longer than needed
                // @ts-ignore
                pad = new UString(padString);
            }
            return pad.slice(0, targetLength) + String(this);
        }
    };
    /**
     * 𠮷 134071 20bb7
     */
    UString.prototype.codePointAt = function (pos) {
        return this.toArray()[pos].codePointAt(0);
    };
    UString.create = function (str) {
        var argv = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            argv[_i - 1] = arguments[_i];
        }
        return new (this.bind.apply(this, __spreadArrays([void 0, str], argv)))();
    };
    Object.defineProperty(UString, "support", {
        /**
         * 顯示 目前支援 unicode 的 method
         */
        get: function () {
            var prototype = es6_class_prototype_1["default"](this);
            return Object.keys(prototype).reduce(function (a, b) {
                if (exports.STRING_PROTOTYPE.includes(b)) {
                    a[b] = true;
                }
                return a;
            }, {});
        },
        enumerable: false,
        configurable: true
    });
    UString.indexOf = function (str, search, start) {
        if (start === void 0) { start = 0; }
        return this.create(str).indexOf(search, start);
    };
    UString.includes = function (str, search, start) {
        if (start === void 0) { start = 0; }
        return this.create(str).includes(search, start);
    };
    /**
     * splits a String object into an array of strings by separating the string into substrings, using a specified separator string to determine where to make each split.
     */
    UString.split = function (str, separator, limit) {
        return this.create(str).split(separator, limit);
    };
    UString.substr = function (str, start, length) {
        return this.create(str).substr(start, length);
    };
    UString.substring = function (str, start, indexEnd) {
        return this.create(str).substring(start, indexEnd);
    };
    UString.slice = function (str, start, indexEnd) {
        return this.create(str).slice(start, indexEnd);
    };
    UString.charAt = function (str, index) {
        return this.create(str).charAt(index);
    };
    UString.padEnd = function (str, targetLength, padString) {
        return this.create(str).padEnd(targetLength, padString);
    };
    UString.padStart = function (str, targetLength, padString) {
        return this.create(str).padStart(targetLength, padString);
    };
    UString.startsWith = function (str, search, pos) {
        return this.create(str).startsWith(search, pos);
    };
    UString.endsWith = function (str, search, length) {
        return this.create(str).endsWith(search, length);
    };
    Object.defineProperty(UString.prototype, "charLength", {
        get: function () {
            return UString.toArray(this).length;
        },
        enumerable: false,
        configurable: true
    });
    UString.prototype.size = function () {
        return UString.toArray(this).length;
    };
    UString.size = function (str) {
        return this.create(str).size();
    };
    /**
     * 𠮷 134071 20bb7
     */
    UString.codePointAt = function (str, pos) {
        return this.create(str).codePointAt(pos);
    };
    UString.UString = UString;
    /**
     * @private
     */
    UString["default"] = UString;
    return UString;
}(String));
exports.UString = UString;
exports["default"] = UString;
// @ts-ignore
Object.defineProperty(UString, "__esModule", { value: true });
// @ts-ignore
UString["default"] = UString.UString = UString;
// @ts-ignore
exports = Object.freeze(exports);
