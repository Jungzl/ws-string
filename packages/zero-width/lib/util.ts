/**
 * Created by user on 2019/8/12.
 */

/**
 * simple merge, not real regexp merge
 * @private
 */
export function _regexpMerge(re: (string | RegExp)[])
{
	return new RegExp('[' + re.map(r => (typeof r === 'string' ? r : r.source).replace(/^\[/, '').replace(/\]$/, ''))
		.join('') + ']', 'ug')
}
