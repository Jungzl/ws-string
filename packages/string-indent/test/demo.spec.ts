//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { inspect } from 'util';
import { detectIndentLine } from '../src/index';

beforeAll(async () =>
{

});

describe(`base`, () =>
{

	([
		['    at pkg\\index.ts:64:6'],
		['\n'],
		[''],
	] as [string, string?][]).forEach(input => {

		test(inspect(input[0]), () =>
		{

			let actual = detectIndentLine(input[1] ?? input[0]);

			expect(actual).toMatchSnapshot();

		});

	});

})

describe(`no-break space`, () =>
{

	([
		['\xa0    at pkg\\index.ts:64:6'],
		['\xa0\n'],
		['\xa0'],
	] as [string, string?][]).forEach(input =>
	{

		test(inspect(input[0]), () =>
		{

			let actual = detectIndentLine(input[1] ?? input[0], {
				includeNoBreakSpace: true,
			});

			expect(actual).toMatchSnapshot();

		});

	});

})
