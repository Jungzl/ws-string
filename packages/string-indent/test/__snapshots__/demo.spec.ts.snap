// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base '    at pkg\\\\index.ts:64:6' 1`] = `
{
  "body": "at pkg\\index.ts:64:6",
  "bool": true,
  "indent": "    ",
  "input": "    at pkg\\index.ts:64:6",
}
`;

exports[`base '' 1`] = `
{
  "body": "",
  "bool": false,
  "indent": "",
  "input": "",
}
`;

exports[`base '\\n' 1`] = `
{
  "body": "
",
  "bool": false,
  "indent": "",
  "input": "
",
}
`;

exports[`no-break space '     at pkg\\\\index.ts:64:6' 1`] = `
{
  "body": "at pkg\\index.ts:64:6",
  "bool": true,
  "indent": "     ",
  "input": "     at pkg\\index.ts:64:6",
}
`;

exports[`no-break space ' ' 1`] = `
{
  "body": "",
  "bool": true,
  "indent": " ",
  "input": " ",
}
`;

exports[`no-break space ' \\n' 1`] = `
{
  "body": "
",
  "bool": true,
  "indent": " ",
  "input": " 
",
}
`;
